File: backend/app/database/models/analytics.py
from __future__ import annotations

from datetime import datetime
from uuid import uuid4

from sqlalchemy import ForeignKey, Numeric, String
from sqlalchemy.orm import Mapped, mapped_column

from app.database.base import Base, TimestampMixin


class PerformanceMetrics(Base, TimestampMixin):
    __tablename__ = "performance_metrics"

    id: Mapped[str] = mapped_column(primary_key=True, default=lambda: str(uuid4()))
    portfolio_id: Mapped[str] = mapped_column(ForeignKey("portfolios.id"), index=True, nullable=False)
    user_id: Mapped[str] = mapped_column(ForeignKey("users.id"), index=True, nullable=False)
    period_start: Mapped[datetime] = mapped_column(index=True, nullable=False)
    period_end: Mapped[datetime] = mapped_column(index=True, nullable=False)
    period_type: Mapped[str] = mapped_column(String(20), nullable=False)

    total_return: Mapped[float] = mapped_column(Numeric(20, 8), nullable=True)
    sharpe_ratio: Mapped[float] = mapped_column(Numeric(10, 6), nullable=True)
    sortino_ratio: Mapped[float] = mapped_column(Numeric(10, 6), nullable=True)
    calmar_ratio: Mapped[float] = mapped_column(Numeric(10, 6), nullable=True)
    profit_factor: Mapped[float] = mapped_column(Numeric(20, 8), nullable=True)
    win_rate: Mapped[float] = mapped_column(Numeric(5, 4), nullable=True)
    expectancy: Mapped[float] = mapped_column(Numeric(20, 8), nullable=True)
    max_drawdown: Mapped[float] = mapped_column(Numeric(20, 8), nullable=True)
    risk_adjusted_return: Mapped[float] = mapped_column(Numeric(20, 8), nullable=True)

    total_trades: Mapped[int] = mapped_column(nullable=True, default=0)
    winning_trades: Mapped[int] = mapped_column(nullable=True, default=0)
    losing_trades: Mapped[int] = mapped_column(nullable=True, default=0)
    gross_profit: Mapped[float] = mapped_column(Numeric(20, 8), nullable=True)
    gross_loss: Mapped[float] = mapped_column(Numeric(20, 8), nullable=True)
    volatility: Mapped[float] = mapped_column(Numeric(20, 8), nullable=True)
    downside_deviation: Mapped[float] = mapped_column(Numeric(20, 8), nullable=True)

    __table_args__ = (
        {"sqlite_autoincrement": True},
    )

File: frontend/src/schemas/analytics.ts
import { z } from "zod";

export const PortfolioAnalyticsRequestSchema = z.object({
  portfolio_id: z.string().uuid().optional(),
  date_from: z.string().date().optional(),
  date_to: z.string().date().optional(),
  risk_free_rate: z.number().default(0),
});

export const PerformanceMetricsSchema = z.object({
  total_return: z.number().nullable().optional(),
  sharpe_ratio: z.number().nullable().optional(),
  sortino_ratio: z.number().nullable().optional(),
  calmar_ratio: z.number().nullable().optional(),
  profit_factor: z.number().nullable().optional(),
  win_rate: z.number().nullable().optional(),
  expectancy: z.number().nullable().optional(),
  max_drawdown: z.number().nullable().optional(),
  risk_adjusted_return: z.number().nullable().optional(),
  period_start: z.string().datetime().nullable().optional(),
  period_end: z.string().datetime().nullable().optional(),
  total_trades: z.number().int().default(0),
  winning_trades: z.number().int().default(0),
  losing_trades: z.number().int().default(0),
  gross_profit: z.number().nullable().optional(),
  gross_loss: z.number().nullable().optional(),
});

export const PortfolioAnalyticsResponseSchema = z.object({
  portfolio: z.object({
    id: z.string(),
    user_id: z.string(),
    portfolio_name: z.string(),
    currency: z.string(),
    total_value: z.number(),
    cash_balance: z.number(),
    risk_score: z.number(),
    created_at: z.string(),
    updated_at: z.string(),
  }).nullable().optional(),
  metrics: PerformanceMetricsSchema,
  chart_data: z.record(z.array(z.record(z.union([z.number(), z.string()])))),
});